Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp

```diff
@@ -199,6 +268,269 @@ namespace AZ
+                }
+
+                // At this point shaderTemplate.m_azsli should be an absolute path due to ResolvePathReference() being called above.
+                // It might be better for the include path to be relative to the generated .shader file path in the intermediate cache,
+                // so the project could be renamed or moved without having to rebuild the cache. But there's a good chance that moving
+                // the project would require a rebuild of the cache anyway.
+                AZStd::string includeTemplateAzslPath = shaderTemplate.m_azsli;
+
+                // Intermediate azsl file
+
+                AZStd::string generatedAzsl = AZStd::string::format(
+                    "// This code was generated by %s. Do not modify.\n"
+                    "#include <%s>\n",
+                    MaterialTypeBuilderName,
+                    includeTemplateAzslPath.c_str());
+
+                // The exact same material-specific shader code is included in every shader
+                if (!materialAzsliFilePath.empty())
+                {
+                    generatedAzsl += AZStd::string::format("#include <%s>\n", materialAzsliFilePath.c_str());
+                }
+
+                // The "_DEFINED" macros allow the material type to provide code for only the functions it needs to customize,
+                // and all other functions will be stubbed out.
+                generatedAzsl +=
+                    " \n"
+                    "#if !MaterialFunction_AdjustVertexData_DEFINED                                                    \n"
+                    "    void MaterialFunction_AdjustVertexData(in float3 positionLS, inout VertexData vertexData) {}  \n"
+                    "    #define MaterialFunction_AdjustVertexData_DEFINED 1                                           \n"
+                    "#endif                                                                                            \n"
+                    "                                                                                                  \n"
+                    "#if !MaterialFunction_AdjustSurface_DEFINED && MATERIALPIPELINE_SHADER_HAS_PIXEL_STAGE            \n"
+                    "    void MaterialFunction_AdjustSurface(inout Surface outSurface) {}                              \n"
+                    "    #define MaterialFunction_AdjustSurface_DEFINED 1                                              \n"
+                    "#endif                                                                                            \n"
+                    ;
+
+                AZ::IO::Path shaderName = shaderTemplate.m_shader;
+                shaderName = shaderName.Filename(); // Removes the folder path
+                shaderName = shaderName.ReplaceExtension(""); // This will remove the ".template" extension
+                shaderName = shaderName.ReplaceExtension(""); // This will remove the ".shader" extension
+
```

Comment:
As we discussed before, I got rid of the template file extension for material canvas templates and it might be worth doing the same here. The purpose of the template file extension was to make the template files easily identifiable while also hiding them from the AP. But hiding them from the AP made it easy for other changes, like include files being moved around, to break the templates. Now, templates use the standard file extension, which allows them to be automatically recognized, processed, and validated by the AP. The template is just defined by references in the output node configuration but files for the current template are also contained in the same folder. Most of the template files are treated as text files that will be scanned for markers that can be filled in or substituted. The material type and shader files can be deserialized directly to make other modifications to the data structures.

Reply:
>But hiding them from the AP made it easy for other changes, like include files being moved around, to break the templates.
This isn't a concern for the material pipeline system because there is an asset builder that uses these files, and will readily fail if any breaking changes are made.